test/unit/matrix/test_copy.cpp

```diff
@@ -0,0 +1,384 @@
+      GlobalElementSize size = globalTestSize(test.size, comm_grid.size());
+
+      Distribution distribution(size, test.block_size, comm_grid.size(), comm_grid.rank(), {0, 0});
+      LayoutInfo layout = tileLayout(distribution.localSize(), test.block_size);
+
+      MemoryViewT mem_src(layout.minMemSize());
+      MatrixT mat_src = createMatrixFromTile<Device::CPU>(size, test.block_size, comm_grid, mem_src());
+      set(mat_src, inputValues<TypeParam>);
+      MatrixConstT mat_src_const = std::move(mat_src);
+
+      GPUMemoryViewT mem_gpu1(layout.minMemSize());
+      GPUMatrixT mat_gpu1 =
+          createMatrixFromTile<Device::GPU>(size, test.block_size, comm_grid, mem_gpu1());
+
+      GPUMemoryViewT mem_gpu2(layout.minMemSize());
+      GPUMatrixT mat_gpu2 =
+          createMatrixFromTile<Device::GPU>(size, test.block_size, comm_grid, mem_gpu2());
+
+      MemoryViewT mem_dst(layout.minMemSize());
+      MatrixT mat_dst = createMatrixFromTile<Device::CPU>(size, test.block_size, comm_grid, mem_dst());
+      set(mat_dst, outputValues<TypeParam>);
+
+      copy(mat_src_const, mat_gpu1);
+      copy(mat_gpu1, mat_gpu2);
+      copy(mat_gpu2, mat_dst);
+
+      CHECK_MATRIX_NEAR(inputValues<TypeParam>, mat_dst, 0, TypeUtilities<TypeParam>::error);
+    }
+  }
+}
+#endif
+
+struct SubMatrixCopyConfig {
+  GlobalElementSize full_in;
+  GlobalElementSize full_out;
+
+  TileElementSize tile_size;
+
+  GlobalElementIndex sub_origin_in;
+  GlobalElementIndex sub_origin_out;
+
+  GlobalElementSize sub_size;
+};
+
+comm::Index2D alignSubRankIndex(const Distribution& dist_in, const GlobalElementIndex& offset_in,
+                                const GlobalElementIndex& offset_out) {
+  const auto pos_mod = [](const auto& a, const auto& b) {
+    const auto mod = a % b;
+    return (mod >= 0) ? mod : (mod + b);
+  };
+
+  const comm::Size2D grid_size = dist_in.commGridSize();
+
+  const comm::Index2D sub_rank(dist_in.template rank_global_element<Coord::Row>(offset_in.row()),
+                               dist_in.template rank_global_element<Coord::Col>(offset_in.col()));
+  const GlobalTileIndex offset_tl(offset_out.row() / dist_in.blockSize().rows(),
+                                  offset_out.col() / dist_in.blockSize().cols());
+
+  return {pos_mod(sub_rank.row() - static_cast<comm::IndexT_MPI>(offset_tl.row()), grid_size.rows()),
+          pos_mod(sub_rank.col() - static_cast<comm::IndexT_MPI>(offset_tl.col()), grid_size.cols())};
+}
+
+bool isFullMatrix(const Distribution& dist_full, const GlobalElementIndex& sub_origin,
+                  const GlobalElementSize& sub_size) noexcept {
+  return sub_origin == GlobalElementIndex{0, 0} && sub_size == dist_full.size();
+}
+
+bool isInSub(const GlobalElementIndex& ij, const GlobalElementIndex& origin,
+             const GlobalElementSize& sub_size) noexcept {
+  return ij.row() >= origin.row() && ij.col() >= origin.col() &&
+         ij.isIn(sizeFromOrigin(origin + sub_size));
```

Comment:
Might be helpful to add this helper to indices directly?

Reply:
see 38f96c9